### DISCUSSION TITLE
2d simulation with variable grid spacings

### DISCUSSION THREAD

**Discussion by parfenyev** [CliMA/Oceananigans.jl#2516] [Computational science!] [answered]
Hello everyone,

I am trying to perform 2d simulation with variable grid spacings in `x-` and `y`-directions. The following is a minimal example that gives the error:
```
using Oceananigans, Statistics, Printf

Nx, Ny = 512, 512;
Lx, Ly = 2π, 2π;

chebychev_spaced_x_faces(i) = Lx/2 - Lx/2 * cos(π * (i - 1) / Nx);
chebychev_spaced_y_faces(j) = Ly/2 - Ly/2 * cos(π * (j - 1) / Ny);

grid = RectilinearGrid(size = (Nx, Ny), 
                              topology=(Periodic, Periodic, Flat),
                              x = chebychev_spaced_x_faces,
                              y = chebychev_spaced_y_faces)

model = NonhydrostaticModel(timestepper = :RungeKutta3,
                              advection = UpwindBiasedFifthOrder(),
                                   grid = grid,
                               buoyancy = nothing,
                                tracers = nothing,
                                closure = ScalarDiffusivity(ν=1e-5))
```
```
MethodError: no method matching PressureSolver(::CPU, ::Oceananigans.Grids.ZRegRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, Float64, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, CPU})
Closest candidates are:
  PressureSolver(::Any, ::RectilinearGrid{<:Any, <:Any, <:Any, <:Any, <:Number, <:Number, <:Number}) at C:\Users\parfe\.julia\packages\Oceananigans\jmNfq\src\Models\NonhydrostaticModels\NonhydrostaticModels.jl:28
  PressureSolver(::Any, ::RectilinearGrid{<:Any, <:Any, <:Any, <:Any, <:Number, <:Number}) at C:\Users\parfe\.julia\packages\Oceananigans\jmNfq\src\Models\NonhydrostaticModels\NonhydrostaticModels.jl:29
  PressureSolver(::Any, ::Oceananigans.ImmersedBoundaries.ImmersedBoundaryGrid) at C:\Users\parfe\.julia\packages\Oceananigans\jmNfq\src\Models\NonhydrostaticModels\NonhydrostaticModels.jl:32

Stacktrace:
 [1] NonhydrostaticModel(; grid::Oceananigans.Grids.ZRegRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, Float64, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, OffsetArrays.OffsetVector{Float64, Vector{Float64}}, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, CPU}, clock::Clock{Float64}, advection::UpwindBiasedFifthOrder, buoyancy::Nothing, coriolis::Nothing, stokes_drift::Nothing, forcing::NamedTuple{(), Tuple{}}, closure::ScalarDiffusivity{Oceananigans.TurbulenceClosures.ExplicitTimeDiscretization, Oceananigans.TurbulenceClosures.ThreeDimensionalFormulation, Float64, Float64}, boundary_conditions::NamedTuple{(), Tuple{}}, tracers::Nothing, timestepper::Symbol, background_fields::NamedTuple{(), Tuple{}}, particles::Nothing, velocities::Nothing, pressures::Nothing, diffusivity_fields::Nothing, pressure_solver::Nothing, immersed_boundary::Nothing, auxiliary_fields::NamedTuple{(), Tuple{}})
   @ Oceananigans.Models.NonhydrostaticModels C:\Users\parfe\.julia\packages\Oceananigans\jmNfq\src\Models\NonhydrostaticModels\nonhydrostatic_model.jl:185
 [2] top-level scope
   @ In[45]:14
 [3] eval
   @ .\boot.jl:373 [inlined]
 [4] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
   @ Base .\loading.jl:1196
```
How can this be fixed?

**Comment by glwagner** [comment, marked as answer]
Variable horizontal grid spacing isn't supported by `NonhydrostaticModel`, because the only pressure solvers that are implemented are fully FFT based (for regular in all directions) and a combined FFT + tridiagonal solve (for regular in horizontal directions, stretched in the vertical).

**Reply by parfenyev** [reply]
Thanks. Maybe you know some other package where this feature is implemented? I would like to simulate 2D turbulence with no-slip boundary conditions and variable grid spacing (to resolve boundary layers).

**Reply by francispoulin** [reply]
@glwagner can we use the hydrostatic model to do this?  

@parfenyev  If you wawnt to study 2D turbulence, you should not get any non-hydrostatic features so I would think that the hydrostatic model would be a good play to try this, if it can handle a cheb grid on both directions.

**Reply by glwagner** [reply]
Yes, you can use the hydrostatic model, though we haven't tried it much in that configuration. But it should use a matrix based solver on a horizontally stretched rectilinear grid for the implicit free surface step, and the dynamics are close to what you'd get from the nonhydrostatic model provided the wave speed is much faster than the dynamics time-scale.

I'd be curious what you find. Unless you're doing a large problem, I think it typically makes sense to simply use high resolution everywhere on a regular grid --- because the FFT-based solver is so fast. The issue is that at relatively low resolution there aren't enough points to achieve significant computational savings by stretching the grid --- even if we neglect the difference in pressure solver speed for regular vs stretched grid.

But if you have a large domain with narrow boundary layers then its possible that regular spacing will be prohibatively expensive. Perhaps 512^2 is big enough to warrant it!

**Reply by glwagner** [reply]
> Maybe you know some other package where this feature is implemented?

Do you mean, other software where one can simulate two-dimensional turbulence problems with horizontally-variable grid spacing? There's probably quite a bit out there.

**Reply by parfenyev** [reply]
> Do you mean, other software where one can simulate two-dimensional turbulence problems with horizontally-variable grid spacing? There's probably quite a bit out there.

Yes, I mean other software. As far as I understand, the hydrostatic model in `Oceananigans.jl` will spend computational time to calculate the shape of the surface -- in my case this is not required at all.

**Reply by glwagner** [reply]
I don't know of any software designed for this particular problem (doubly-bounded two-dimensional simulations on stretched grids --- there's lots of spectral-based software for uniform grids though). I'm sure people would love to know about different options if you find them so that'd be good to post here! [dedalus](https://dedalus-project.org) is worth checking out but I believe can only do a Chebyshev grid in one direction, not two (though this may have or could change at any time so worth reading their documentation).

As for the software in this repository (Oceananigans), if the hydrostatic model used in this configuration, g * eta essentially takes the form of "pressure", so the actual computations performed are extremely similar. The main difference is that w must be "integrated" (a tiny computation in this 2D domain). The solvers though are actually quite similar. In fact, if using an iterative solver and a non-zero "Mach" number (the ratio u / sqrt(g *h)); sqrt(g * h) is the surface wave speed then a free surface is _more_ efficient, because the iterations converge more quickly. However, care must be taken because the physics is identical strictly only in the limit u / c -> 0.

Here's a little demonstration:

```julia
using Oceananigans

grid = RectilinearGrid(CPU(), size=(128, 128), x=(0, 2π), y=(0, 2π), topology=(Periodic, Periodic, Flat))
nonhydrostatic_model = NonhydrostaticModel(; grid, advection=WENO5())

grid = RectilinearGrid(CPU(), size=(128, 128, 1), x=(0, 2π), y=(0, 2π), z=(0, 1), topology=(Periodic, Periodic, Bounded))
hydrostatic_model = HydrostaticFreeSurfaceModel(; grid,
                                                momentum_advection = WENO5(),
                                                free_surface = ImplicitFreeSurface(gravitational_acceleration=10),
                                                tracers = nothing,
                                                buoyancy = nothing)

for model in (nonhydrostatic_model, hydrostatic_model)
    ϵ(x, y, z) = 2rand() - 1
    set!(model, u=ϵ, v=ϵ)
    simulation = Simulation(model; Δt=1e-4, stop_iteration=10)
    run!(simulation)
    simulation.stop_iteration += 100

    start = time_ns()
    run!(simulation)
    finish = time_ns()
    elapsed = 1e-9 * (finish - start)

    @info """ $(summary(model))

    took

    $(prettytime(elapsed))
    """
end
```

which produces

```
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (173.916 μs)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (4.162 ms).
[ Info: Simulation is stopping. Model iteration 10 has hit or exceeded simulation stop iteration 10.
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (110.625 μs)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (3.839 ms).
[ Info: Simulation is stopping. Model iteration 110 has hit or exceeded simulation stop iteration 110.
┌ Info:  NonhydrostaticModel{CPU, RectilinearGrid}(time = 1.100 seconds, iteration = 110)
│
│ took
│
└ 377.874 ms
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (762.416 μs)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (5.091 ms).
[ Info: Simulation is stopping. Model iteration 10 has hit or exceeded simulation stop iteration 10.
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (707.167 μs)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (4.050 ms).
[ Info: Simulation is stopping. Model iteration 110 has hit or exceeded simulation stop iteration 110.
┌ Info:  HydrostaticFreeSurfaceModel{CPU, RectilinearGrid}(time = 1.100 seconds, iteration = 110)
│
│ took
│
└ 389.103 ms
```

nearly identical run times for this small problem. For 512^2, the times are 5.64 and 5.04 s for the nonhydrostatic and hydrostatic model (ie the hydrostatic model is slightly faster, at least for that run).

On a regular grid, both models use FFT-based solvers for either pressure or free surface (for either nonhydrostatic or hydrostatic mode).

**Reply by glwagner** [reply]
Here's another comparison between regular and stretched grids:

```julia
using Oceananigans

Lx = Ly = 2π
Nx = Ny = 512

chebychev_spaced_x_faces(i) = Lx/2 - Lx/2 * cos(π * (i - 1) / Nx);
chebychev_spaced_y_faces(j) = Ly/2 - Ly/2 * cos(π * (j - 1) / Ny);

stretched_grid = RectilinearGrid(size = (Nx, Ny, 1),
                                 topology=(Bounded, Bounded, Bounded),
                                 x = chebychev_spaced_x_faces,
                                 y = chebychev_spaced_y_faces,
                                 z = (0, 1))

regular_grid = RectilinearGrid(size = (Nx, Ny, 1),
                               topology=(Bounded, Bounded, Bounded),
                               x = (0, 2π),
                               y = (0, 2π),
                               z = (0, 1))


for grid in (stretched_grid, regular_grid)

    model = HydrostaticFreeSurfaceModel(; grid,
                                        momentum_advection = WENO5(; grid),
                                        free_surface = ImplicitFreeSurface(gravitational_acceleration=10),
                                        tracers = nothing,
                                        buoyancy = nothing)

    ϵ(x, y, z) = 2rand() - 1
    set!(model, u=ϵ, v=ϵ)
    simulation = Simulation(model; Δt=1e-4, stop_iteration=10)
    run!(simulation)
    simulation.stop_iteration += 100

    start = time_ns()
    run!(simulation)
    finish = time_ns()
    elapsed = 1e-9 * (finish - start)

    @info """ A hydrostatic model with

    $(model.grid)

    took

    $(prettytime(elapsed))
    """
end
```

produces

```
┌ Info:  A hydrostatic model with
│
│ 512×512×1 RectilinearGrid{Float64, Bounded, Bounded, Bounded} on CPU with 3×3×3 halo
│ ├── Bounded  x ∈ [0.0, 6.28319] variably spaced with min(Δx)=5.91396e-5, max(Δx)=0.0192765
│ ├── Bounded  y ∈ [0.0, 6.28319] variably spaced with min(Δy)=5.91396e-5, max(Δy)=0.0192765
│ └── Bounded  z ∈ [0.0, 1.0]     regularly spaced with Δz=1.0
│
│ took
│
└ 17.053 seconds
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (5.532 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (58.431 ms).
[ Info: Simulation is stopping. Model iteration 10 has hit or exceeded simulation stop iteration 10.
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (4.151 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (52.953 ms).
x[ Info: Simulation is stopping. Model iteration 110 has hit or exceeded simulation stop iteration 110.
┌ Info:  A hydrostatic model with
│
│ 512×512×1 RectilinearGrid{Float64, Bounded, Bounded, Bounded} on CPU with 3×3×3 halo
│ ├── Bounded  x ∈ [-3.2953e-18, 6.28319] regularly spaced with Δx=0.0122718
│ ├── Bounded  y ∈ [-3.2953e-18, 6.28319] regularly spaced with Δy=0.0122718
│ └── Bounded  z ∈ [0.0, 1.0]             regularly spaced with Δz=1.0
│
│ took
│
└ 5.217 seconds
```

so the regular grid is about 3x faster at the same resolution. In this configuration, a regular grid with 1024^2 resolution runs in about the same time as a 512^2 stretched grid with this extreme stretching (factor of 1e-3 variation in grid spacing)

There's another issue with this setup though. The extremely small grid spacing near the boundaries means that the diffusive CFL constraint might be very strict for this model (depending on the viscosity). Resolving that issue probably requires treating diffusion implicitly, which we don't support (we only support implicit diffusion in the vertical direction). Implicit treatment of diffusion on a stretched grid also requires an iterative solver of some kind.

**Reply by francispoulin** [reply]
Very nice result!  The fact that the cheb grid is only a few orders of magnitude slower is rather impressive.  In my experience cheb grids tend to be a lot slower because the grid spacing near the boundary is O(dx^2), so the CFL condition is much more restrictive.  Here you are able to use a reasonable time step for both.

**Reply by glwagner** [reply]
True, but there's no closure here! I've only compared cost per time-step, not time-to-solution (effectively). I think if we have explicit diffusion we would indeed require a very small time-step, thus exacerbating the problem of a slow time-step further. One recourse in the finite volume context is an iterative solver for implicit diffusion, which would again increase the computational expense, likely by another factor of 3 in this two-dimensional problem (but much more in three-dimensions). Numerically this is not all that challenging, but implementing it would be a substantial extension of existing functionality.

**Reply by glwagner** [reply]
Exasim might be able to handle this problem --- it treates the entire tendency implicitly using sophisticated general iterative solvers (that work for nonlinear and linear terms!), so is probably well suited to problems with extreme grid stretching like this one: https://github.com/exapde/Exasim

**Reply by parfenyev** [reply]
Thank you very much! I have found that people use pseudospectral methods based on Chebyshev polynomials to solve such problems, see e.g. [this paper](https://www.sciencedirect.com/science/article/pii/S0021999197957995). Can you comment on whether such a method will give a significant increase in the speed of calculations in comparison with the hydrostatic model from `Oceananigans.jl`?

**Reply by glwagner** [reply]
I don't know. I think double Chebyshev with implicit diffusion could be a "good choice" for some applications --- like two-dimensional, doubly-bounded Navier-Stokes simulations that resolve viscous boundary layers but have high Reynolds number in the interior. But I'm not an expert.

**Reply by francispoulin** [reply]
I also can't say much but it depends very much on what you want to do.  If you want to resolve boundary layers then perhaps a cheb grid is the way to go, but it will always be more expensive.

**Comment by glwagner** [comment]
```julia
using Oceananigans, Statistics, Printf

Nx, Ny = 512, 512;
Lx, Ly = 2π, 2π;

chebychev_spaced_x_faces(i) = Lx/2 - Lx/2 * cos(π * (i - 1) / Nx);
chebychev_spaced_y_faces(j) = Ly/2 - Ly/2 * cos(π * (j - 1) / Ny);

grid = RectilinearGrid(size = (Nx, Ny, 1), 
                       topology=(Bounded, Bounded, Bounded),
                       # topology=(Periodic, Periodic, Bounded),
                       x = chebychev_spaced_x_faces,
                       y = chebychev_spaced_y_faces,
                       z = (0, 1))

model = HydrostaticFreeSurfaceModel(; grid,
                                    momentum_advection = UpwindBiasedFifthOrder(),
                                    buoyancy = nothing,
                                    tracers = nothing,
                                    free_surface = ImplicitFreeSurface(gravitational_acceleration=10), # c = sqrt(10)
                                    closure = ScalarDiffusivity(ν=1e-3))

ϵ(x, y, z) = 2rand() - 1
set!(model, u=ϵ, v=ϵ)

simulation = Simulation(model, Δt=0.01, stop_time=10)
run!(simulation)
```

might work

**Reply by francispoulin** [reply]
I tried changing the model as follows and it didn't complain, but I also didn't go any further.

```
 model = HydrostaticFreeSurfaceModel(      
                                                     grid = grid,
                                              buoyancy = nothing,
                                               tracers = nothing,
                                               closure = ScalarDiffusivity(ν=1e-5))
HydrostaticFreeSurfaceModel{CPU, RectilinearGrid}(time = 0 seconds, iteration = 0)
├── grid: 512×512×1 RectilinearGrid{Float64, Periodic, Periodic, Flat} on CPU with 1×1×0 halo
├── timestepper: QuasiAdamsBashforth2TimeStepper
├── tracers: ()
├── closure: ScalarDiffusivity{ExplicitTimeDiscretization}(ν=1.0e-5, κ=NamedTuple())
├── buoyancy: Nothing
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing
```

**Reply by glwagner** [reply]
Do we really want horizontally-periodic with variable spacing? Or do we want `topology = (Bounded, Bounded, Bounded)`?

**Comment by francispoulin** [comment]
I agree that if it's doubly periodic then using a cheb grid might not be very efficient  

I would think that `topology = (Bounded, Bounded, Flat)` would make sense and then you could get boundary layers forming, which is maybe one of the motivations for looking at this?

**Reply by glwagner** [reply]
We can't do `Flat` in the vertical with a free surface (just because of the way the algorithm is written). It might not be hard to do that refactor though.

**Reply by francispoulin** [reply]
Sorry, my mistakes.  Thanks for clarifying.

**Reply by francispoulin** [reply]
Using `topology = (Bounded, Bounded, Flat)`, I was able to create a `ShallowWaterModel` (see below).  This would be maybe a simpler way to study turbulence as there is no inversion to worry about so it should not be too slow.  There is only the `CFL` conditon that could slow things down.

However we have no closer yet.  Maybe we could talk about creating closure for this model @glwagner ?

```
model = ShallowWaterModel(       grid = grid,
                                        gravitational_acceleration = 1.0)
ShallowWaterModel{typename(CPU), Float64}(time = 0 seconds, iteration = 0) 
├── grid: 512×512×1 RectilinearGrid{Float64, Bounded, Bounded, Flat} on CPU with 3×3×0 halo
├── tracers: ()
└── coriolis: Nothing
```

**Reply by glwagner** [reply]
We can talk! I may not be able to devote much time to it right now. I think the shallow water model needs a few more things to make this viable: fixing advection operators, and implicit treatment of the h-equation (so we don't have to resolve wave speed).

That said I actually think the shallow water model is a great test bed for idealized stuff, because its simple (and we should keep it that way). We could develop rigid lid, linear-h and implicit-h capabilities and then it might be even slightly more optimized than `NonhydrostaticModel` for a range of 2D problems.

**Reply by francispoulin** [reply]
All great stuff and happy to talk when you have time, but know you are very busy on **many** other fronts right now
